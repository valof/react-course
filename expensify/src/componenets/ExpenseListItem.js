import React from 'react';
import { NavLink } from 'react-router-dom'


const ExpenseListItem = (props) => (
  <div >
    <NavLink to={'/edit/'+props.expense.id} className={({ isActive }) => (isActive ? 'is-active' : 'inactive')}>
      <h3>This is my Expense # {props.index+1}:</h3>
    </NavLink>
    <p/>
      Description: {props.expense.description}
    <p/>
      Amount: {props.expense.amount}
    <p/>
      CreatedAt: {props.expense.createdAt}
    <p/>
  </div >
);
export default ExpenseListItem;