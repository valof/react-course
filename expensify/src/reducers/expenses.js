import { createSlice} from '@reduxjs/toolkit'

const initialExpenseDefaultState = [];

export const sliceExpense = createSlice({
    name: 'expenses',
    initialState: initialExpenseDefaultState, // Define initial state
    reducers: {
      addExpense: (state, action) =>  [...state ,action.payload],
      removeExpense: (state, action) => state.filter( elem => elem.id != action.payload.id ),
      editExpense: (state, action) => state.map((expense) => {
        if (action.payload.id === expense.id) {
          return {
            ...expense,
            ...action.payload
          }
        } else {
          return expense;
        }
      })
    },
  })


export const reducerExpense=sliceExpense.reducer;
