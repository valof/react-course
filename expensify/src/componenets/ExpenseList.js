import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/selector'

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses!</p>
             ) : (
                props.expenses.map((expense, index) => (
                    <ExpenseListItem key = {expense.id} expense={expense} index = {index}/>
                ))
             )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.reducer.espenses, state.reducer.filters),
        filters: state.reducer.filters
    };
};


export default connect(mapStateToProps)(ExpenseList);
