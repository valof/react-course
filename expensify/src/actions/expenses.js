import {sliceExpense} from '../reducers/expenses';

export const {addExpense, removeExpense, editExpense} = sliceExpense.actions;
