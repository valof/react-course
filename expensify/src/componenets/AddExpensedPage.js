import React from 'react';
import ExpenseForm from './ExpenseForm'
import {addExpense, removeExpense, editExpense}  from '../actions/expenses';
import {useNavigate} from "react-router-dom"
import {useDispatch, connect} from "react-redux"

const withRouter = WrappedComponent => props => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  // other hooks

  return (
    <WrappedComponent
      {...props}
      {...{ navigate, dispatch }}
    />
  );
};

class AddExpensePage extends React.Component {

onSubmit = (expense) => {
  this.props.onSubmit(expense);
  this.props.navigate("/")
}

render () {
    return (
      <div>
        <h1>Expense Form</h1>
        <ExpenseForm  
          onSubmit={this.onSubmit}
        />
      </div>
      );
  }
};

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (expense) => dispatch (addExpense(expense))
})

export const AddExpenseDisconnected = withRouter(AddExpensePage);

export default connect(undefined, mapDispatchToProps)(AddExpenseDisconnected);


// This is how it was:
//
// import React from 'react';
// import ExpenseForm from './ExpenseForm'
// import {addExpense, removeExpense, editExpense}  from '../actions/expenses';
// import {useNavigate} from "react-router-dom"
// import {useDispatch} from "react-redux"


// const AddExpensePage = (props) => {
//   const navigate = useNavigate();
//   const dispatch = useDispatch();
//   return (
//   <div>
//     <h1>Expense Form</h1>
//     <ExpenseForm  
//       onSubmit={(p)=>{
//         dispatch(addExpense(p));
//         navigate('/');
//       }}
//     />
//   </div>
//   );
// };

// export default AddExpensePage;