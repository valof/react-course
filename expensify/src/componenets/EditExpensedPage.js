import React from 'react';
import { useParams, useNavigate} from 'react-router-dom';
import {connect} from 'react-redux'
import {addExpense, removeExpense, editExpense}  from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

let id;

const EditExpensedPage = (props) => {
    let { id } = useParams();
    const navigate = useNavigate();
    return (
            <div>
                <ExpenseForm
                    expense = {props.expense}
                    onSubmit = {(p) => {
                        props.dispatch(editExpense({...p, id: id}));
                        console.log('updated', p) ;
                        navigate('/');

                    }}
                />
                <button onClick={() => {
                    props.dispatch(removeExpense({id:props.expense.id}))
                    navigate('/');
                }}>Remove</button>
            </div>
        );
}



const mapStateToProps = (state) => {
    if ( !id ){
        const x = useParams();
        id = x.id;
    }
    const exp = state.reducer.espenses.find((it) => {
        return id === it.id;
    })
    return {
        expense: exp
    };
};

export default connect(mapStateToProps)(EditExpensedPage);
