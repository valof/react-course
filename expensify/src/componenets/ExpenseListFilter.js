import React from 'react';
import { connect } from 'react-redux'
import {setText, sortBy, setStartDate, setEndDate} from '../actions/filters'

const ExpenseFilterList = (props) => (
    <div>
        <input type='text' value={props.filters.text} onChange={(e) => {
            props.dispatch(setText(e.target.value))
        }} />
        <select value={props.filters.sortBy} onChange={(e) => {
            props.dispatch(sortBy(e.target.value))
        }} >
            <option value="date">Date</option>
            <option value="amount">Amount</option>
        </select>
    </div>
);

const mapStateToProps = (state) => {
    return {
        filters: state.reducer.filters
    };
};

export default connect(mapStateToProps)(ExpenseFilterList);