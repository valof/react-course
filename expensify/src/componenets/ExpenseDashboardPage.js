import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilter x = "10" />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;
