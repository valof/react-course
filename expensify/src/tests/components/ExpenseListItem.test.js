import React from 'react'
import { MemoryRouter } from "react-router-dom";
import TestRenderer from 'react-test-renderer';
import {ExpenseListItem} from '../../componenets/ExpenseListItem'
import expenses from '../fixtures/expenses'

test ('should render ExportList with expenses', () => {
    const testRenderer = TestRenderer.create(
        <MemoryRouter>
            <ExpenseListItem expense={expenses[0]} index = {0}/>
        </MemoryRouter>
      );
    expect(testRenderer.toJSON()).toMatchSnapshot();
})
