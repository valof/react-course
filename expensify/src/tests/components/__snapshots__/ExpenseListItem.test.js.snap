// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExportList with expenses 1`] = `
<div>
  <a
    className="inactive"
    href="/edit/1"
    onClick={[Function]}
  >
    <h3>
      This is my Expense # 
      1
      :
    </h3>
  </a>
  <p />
  Description: 
  Mortgage
  <p />
  Amount: 
  505055
  <p />
  CreatedAt: 
  0
  <p />
</div>
`;
